### ISSUE TITLE
Support HTTP signature verifaction from remote instances

### ISSUE THREAD

**Issue by nedjo** [swentel/activitypub#18]
Presumably we need to provide per-Actor public/private key pairs. Possibly relevant: Drupal contrib [Key module](https://www.drupal.org/project/key).

**Comment by swentel** [comment]
Yep, I'm using the key module as well for IndieAuth in indieweb, it's relatively easy to setup.

I know that bridgy fed generates keys per user, which is probably the most secure setup.

(This is something I'm still not sure of in multiple users IndieWeb context for IndieAuth for the JWT token public/private key as the uid is used to generate the token, something I need to verify with the security team :))

**Comment by swentel** [comment]
Actually, I do not use that module at all for the indieweb keys - I completely missed that (I use it for the TFA module). So if you need inspiration for generating keys and not relying on any other contrib modules, see https://github.com/swentel/indieweb/blob/master/modules/indieweb_indieauth/src/IndieAuthClient/IndieAuthClient.php#L119

**Comment by nedjo** [comment]
>  So if you need inspiration for generating keys and not relying on any other contrib modules, see  ...

Nice! In the case that a site had both indieweb and activitypub installed, would we want the same key pair? Should we consider pulling that into a helper module that both indieweb and activitypub could use?

**Comment by swentel** [comment]
Not sure if they should use the same keys, although I'm not an expert on that matter. 

I'm not going to introduce it for the IndieWeb 8.x-1.x branch as that's really too complex for an upgrade path. But it's probably a good a idea for the 8.x-2.x branch. However, instead of creating a new helper module, it probably just makes more sense to start depending on the key module (which has a few dependencies of its own) so at least we're not reinventing the wheel - and it has a really nice plugin system so users can choose which type of keys they want to use.

**Comment by ekes** [comment]
So far in D7 I just used openssl_pkey* functions to generate keypairs and store them in fields on the actor entity. Intention was then to allow use of one of the encrypt modules for the private key field, for additional layer of security. They keys are then used in the code from dgwynne/php-http-signature to sign outgoing messages and verify incoming ones. The code in 99designs/http-signatures-php however looks much nicer for Symfony Drupal 8 integration. 

The only additional brain ache here is that in addition to http signatures, implementations also use linked data signatures. I'd been looking at the code of frendica https://github.com/friendica/friendica/blob/develop/src/Protocol/ActivityPub/Receiver.php to even start getting my head round that.
